src/object_store/src/object/mod.rs

```diff
@@ -318,82 +320,92 @@ impl MonitoredStreamingUploader {
             .observe(data_len as f64);
         let _timer = self
             .object_store_metrics
             .operation_latency
-            .with_label_values(&[self.media_type, operation_type])
+            .with_label_values(&[self.media_type, operation_type_str])
             .start_timer();
         self.operation_size += data_len;
 
         let future = async {
             self.inner
                 .write_bytes(data)
-                .verbose_instrument_await("object_store_streaming_upload_write_bytes")
+                .verbose_instrument_await(operation_type_str)
                 .await
         };
         let res = match self.streaming_upload_timeout.as_ref() {
             None => future.await,
             Some(timeout) => tokio::time::timeout(*timeout, future)
                 .await
                 .unwrap_or_else(|_| {
-                    Err(ObjectError::internal(
-                        "streaming_upload write_bytes timeout",
-                    ))
+                    Err(ObjectError::timeout(format!(
+                        "{} timeout",
+                        operation_type_str
+                    )))
                 }),
         };
 
-        try_update_failure_metric(&self.object_store_metrics, &res, operation_type);
+        try_update_failure_metric(&self.object_store_metrics, &res, operation_type_str);
         res
     }
 
     pub async fn finish(self) -> ObjectResult<()> {
-        let operation_type = "streaming_upload_finish";
+        let operation_type = OperationType::StreamingUploadFinish;
+        let operation_type_str = operation_type.as_str();
         self.object_store_metrics
             .operation_size
-            .with_label_values(&["streaming_upload"])
+            .with_label_values(&[operation_type_str])
             .observe(self.operation_size as f64);
         let _timer = self
             .object_store_metrics
             .operation_latency
-            .with_label_values(&[self.media_type, operation_type])
+            .with_label_values(&[self.media_type, operation_type_str])
             .start_timer();
 
         let future = async {
             self.inner
                 .finish()
-                .verbose_instrument_await("object_store_streaming_upload_finish")
+                .verbose_instrument_await(operation_type_str)
                 .await
         };
-        let res = match self.streaming_upload_timeout.as_ref() {
+        let res = match self.streaming_upload_finish_timeout.as_ref() {
             None => future.await,
             Some(timeout) => tokio::time::timeout(*timeout, future)
                 .await
-                .unwrap_or_else(|_| Err(ObjectError::internal("streaming_upload finish timeout"))),
+                .unwrap_or_else(|_| {
+                    Err(ObjectError::timeout(format!(
+                        "{} timeout",
+                        operation_type_str
+                    )))
+                }),
         };
```

Comment:
https://github.com/risingwavelabs/risingwave/pull/16231#discussion_r1570656358. I think we can deprecate this and ensure all streaming uploader related timeout is maintained in the specific implementaion.

Reply:
Would you explain why is maintaining timeout in the specific implementation preferred? Isn't this PR trying to handle timeout in upper layer as much as possible?